tests/Aspire.Hosting.Azure.Tests/AzureUserAssignedIdentityTests.cs

```diff
@@ -90,12 +90,247 @@ public async Task AddAzureUserAssignedIdentity_WithRoleAssignments_Works()
         var identityRoleAssignments = Assert.Single(model.Resources.OfType<AzureProvisioningResource>(), r => r.Name == "myidentity-roles-myregistry");
         var (_, identityRoleAssignmentsBicep) = await GetManifestWithBicep(identityRoleAssignments, skipPreparer: true);
 
-        Target[] targets = [
-            new Target("bicep", identityBicep),
-            new Target("bicep", registryBicep),
-            new Target("bicep", identityRoleAssignmentsBicep)
-        ];
+        await Verify(identityBicep, "bicep")
+            .AppendContentAsFile(registryBicep, "bicep")
+            .AppendContentAsFile(identityRoleAssignmentsBicep, "bicep");
+    }
+
+    [Fact]
+    public async Task WithAzureUserAssignedIdentity_Works()
+    {
+        // Arrange
+        var builder = TestDistributedApplicationBuilder.Create(DistributedApplicationOperation.Publish);
+
+        var env = builder.AddAzureContainerAppEnvironment("cae");
+
+        var identity = builder.AddAzureUserAssignedIdentity("myidentity");
+
+        // Use generic AddProject with TestProject type
+        var projectBuilder = builder.AddProject<TestProject>("myapp", launchProfileName: null);
+        projectBuilder.WithAzureUserAssignedIdentity(identity);
+
+        using var app = builder.Build();
+        await ExecuteBeforeStartHooksAsync(app, default);
+
+        var model = app.Services.GetRequiredService<DistributedApplicationModel>();
+
+        // Check that only one AzureUserAssignedIdentityResource is created, the one that we explicitly constructed
+        var identityResource = Assert.Single(model.Resources.OfType<AzureUserAssignedIdentityResource>());
+        Assert.Equal("myidentity", identityResource.Name);
+
+        // Check for IComputeResource having the correct identity
+        var computeResource = Assert.Single(model.Resources.OfType<IComputeResource>(), r => r.Name == "myapp");
+        var identityAnnotation = Assert.Single(computeResource.Annotations.OfType<AppIdentityAnnotation>());
+        Assert.Same(identity.Resource, identityAnnotation.IdentityResource);
+        var deploymentTarget = Assert.Single(computeResource.Annotations.OfType<DeploymentTargetAnnotation>());
+
+        var (_, computeResourceBicep) = await GetManifestWithBicep(deploymentTarget.DeploymentTarget, skipPreparer: true);
+
+        await Verify(computeResourceBicep, extension: "bicep");
+    }
+
+    [Fact]
+    public async Task WithAzureUserAssignedIdentity_WithRoleAssignments_Works()
+    {
+        // Arrange
+        var builder = TestDistributedApplicationBuilder.Create(DistributedApplicationOperation.Publish);
+
+        var env = builder.AddAzureContainerAppEnvironment("cae");
+
+        // Use Azure Storage instead of Container Registry
```

Comment:
What is this comment for?

Reply:
I was calling this out as distinct from the Container Registry situation where you associated things with the `WithContainerRegistry` method.

Reply:
```suggestion
        // Use Azure Storage instead of Container Registry to test role assignments on WithReference
```